src/FSharp.Control.TaskSeq/TaskSeqInternal.fs

```diff
@@ -62,7 +62,21 @@ module internal TaskSeqInternal =
 
     let inline raiseEmptySeq () = invalidArg "source" "The input task sequence was empty."
 
-    let inline raiseCannotBeNegative name = invalidArg name "The value must be non-negative."
+    /// Moves the enumerator to its first element, assuming it has just been allocated.
+    /// Raises "The input sequence was empty" if there was no first element.
+    let inline moveFirstOrRaiseUnsafe (e: IAsyncEnumerator<_>) = task {
+        let! hasFirst = e.MoveNextAsync()
+
+        if not hasFirst then
+            invalidArg "source" "The input task sequence was empty."
+    }
+
+    /// Tests the given integer value and raises if it is -1 or lower.
+    let inline raiseCannotBeNegative name value =
+        if value >= 0 then
+            ()
+        else
+            invalidArg name $"The value must be non-negative, but was {value}."
```

Comment:
argumentoufofrange? I guess you're refactoring, which answers that...

Reply:
I'll double-check, but the errors raised should be the same as for similar functions in `Seq.xxx`. I vaguely remember being surprised about this previously. It may be because the argument check is immediate, and the range check happens while iterating the sequence (which, in turn, may have a dynamic length).

Reply:
I actually started the refactoring to limit some allocations in an upcoming PR (and more generally, to improve the code).

Reply:
Forgot about this discussion point. These should be `invalidArg`. We throw `ArgumentOutOfRange` if we find that, while iterating, we end up past the last index, for instance. Here, it is clearly an argument error, as there's never an index that's negative.

(this is my interpretation from MS's logic in the same/similar cases for Array/Seq/List)